File: 39-CombinationSum/39-CombinationSum.java
// Last updated: 8/16/2026, 9:53:23 PM
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        // DFS to get all combinations

        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<>();
        helper(candidates, target, 0, res, new ArrayList<>());
        return res;
    }

    // get a combination with sum = target, 
    // candidates starts from nums[idx]
    // avaiable candidates are nums[idx: ]
    private void helper(
        int[] nums, int target, int idx,
        List<List<Integer>> res, 
        List<Integer> comb) {
        if (target < 0) return;
        if (target == 0) {
            // System.out.println(comb);
            res.add(new ArrayList(comb)); // make deep copy of 'comb'
            // System.out.println(res);
            return;
        }

        // iterate all possible candidates
        for (int i = idx; i < nums.length; i ++) {
            comb.add(nums[i]); // add candidate
            helper(nums, target - nums[i], i, res, comb);
            comb.remove(comb.size() - 1); // remove just added candidate
        }

        
    }
}

File: 39-CombinationSum/39-CombinationSum.py
# Last updated: 8/19/2026, 3:10:42 PM
1class Solution:
2    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
3        # backtracking to return ALL possible answers
4        res = []
5        path = []
6
7        def backtrack(start: int, remaining: int):
8            # find combination sum = remaining in candidates[start:]
9            if remaining == 0:
10                res.append(path[:]) # copy path
11                return
12            if remaining < 0:
13                return
14            
15            for i in range(start, len(candidates)):
16                path.append(candidates[i])          # add to curr path
17                backtrack(i, remaining - candidates[i]) # start from i, cause same candidate can be used multiple times
18                path.pop()                          # cancel the adding
19
20        
21        backtrack(0, target)
22
23        return res